src/emulator/functionsRuntimeWorker.ts

```diff
@@ -0,0 +1,210 @@
+    this.id = uuid.v4();
+    this.triggerId = triggerId;
+    this.runtime = runtime;
+
+    this.runtime.events.on("log", (log: EmulatorLog) => {
+      if (log.type === "runtime-status") {
+        if (log.data.state === "idle") {
+          if (this.state === RuntimeWorkerState.BUSY) {
+            this.state = RuntimeWorkerState.IDLE;
+          } else if (this.state === RuntimeWorkerState.FINISHING) {
+            this.log(`IDLE --> FINISHING`);
+            this.runtime.shutdown();
+          }
+        }
+      }
+    });
+
+    this.runtime.exit.then(() => {
+      this.log("exited");
+      this.state = RuntimeWorkerState.FINISHED;
+    });
+  }
+
+  async execute(frb: FunctionsRuntimeBundle, serializedTriggers?: string) {
+    this.state = RuntimeWorkerState.BUSY;
+    const args: FunctionsRuntimeArgs = { frb, serializedTriggers };
+    this.runtime.send(args);
+  }
+
+  get state(): RuntimeWorkerState {
+    return this._state;
+  }
+
+  set state(state: RuntimeWorkerState) {
+    if (state === RuntimeWorkerState.IDLE) {
+      // Remove all temporary log listeners every time we move to IDLE
+      for (const l of this.logListeners) {
+        this.runtime.events.removeListener("log", l);
+      }
+      this.logListeners = [];
+    }
+
+    if (state === RuntimeWorkerState.FINISHED) {
+      this.runtime.events.removeAllListeners();
+    }
+
+    this.log(state);
+    this._state = state;
+    this.stateEvents.emit(this._state);
+  }
+
+  onLogs(listener: LogListener, forever: boolean = false) {
+    if (!forever) {
+      this.logListeners.push(listener);
+    }
+
+    this.runtime.events.on("log", listener);
+  }
+
+  waitForDone(): Promise<any> {
+    if (this.state === RuntimeWorkerState.IDLE || this.state === RuntimeWorkerState.FINISHED) {
+      return Promise.resolve();
+    }
+
+    return new Promise((res) => {
+      const listener = () => {
+        this.stateEvents.removeListener(RuntimeWorkerState.IDLE, listener);
+        this.stateEvents.removeListener(RuntimeWorkerState.FINISHED, listener);
+        res();
+      };
+
+      // Finish on either IDLE or FINISHED states
+      this.stateEvents.once(RuntimeWorkerState.IDLE, listener);
+      this.stateEvents.once(RuntimeWorkerState.FINISHED, listener);
+    });
+  }
+
+  waitForSystemLog(filter: (el: EmulatorLog) => boolean): Promise<EmulatorLog> {
+    return EmulatorLog.waitForLog(this.runtime.events, "SYSTEM", "runtime-status", filter);
+  }
+
+  private log(msg: string): void {
+    EmulatorLogger.log("DEBUG", `[worker-${this.triggerId}-${this.id}]: ${msg}`);
+  }
+}
+
+export class RuntimeWorkerPool {
+  workers: { [triggerId: string]: Array<RuntimeWorker> } = {};
+
+  /**
+   * When code changes (or in some other rare circumstances) we need to get
+   * a new pool of workers. For each IDLE worker we kill it immediately. For
+   * each BUSY worker we move it to the FINISHING state so that it will
+   * kill itself after the next run.
```

Comment:
after it's done with its current task, right?

Reply:
Yep, good catch